### ISSUE TITLE
perf: main window 10–15 fps — GUI thread bound on Qt rhiFlush backing-store composite (multi-panadapter)

### ISSUE THREAD

**Issue by ten9876** [aethersdr/AetherSDR#3797]
## Summary

The main window runs at **10–15 fps** with multiple panadapters open, despite high-end hardware. CPU profiling shows the app is **single-threaded-bound on the GUI thread** in Qt's RHI backing-store composite path — **not** in AetherSDR's own logic (only ~2% of samples) and **not** GPU-starved. This issue captures the profile; the chosen fix is gated on an in-flight investigation of Qt's `rhiFlush` backing-store behavior.

## Repro / environment

- **Radio:** FlexRadio FLEX-8600, **4 panadapters** (4 slices), 20m.
- **Build:** `RelWithDebInfo` (`-O2 -g -DNDEBUG`), current `main` (post-v26.6.4).
- **RHI backend:** **OpenGL** (`SpectrumWidget: QRhi initialized, backend: OpenGL`). On Linux we never call `QRhiWidget::setApi()`, so it defaults to OpenGL.
- **Hardware:** i9-13980HX (32 threads), RTX 4090 Laptop + Intel UHD (hybrid), KDE **Wayland**. Process holds NVIDIA render-node fds → rendering on the 4090, GPU not starved.
- **Symptom:** ~10–15 fps on the main window (measured via MangoHud).

## Profile (perf, DWARF call-graph, main thread, ~18k samples / 6s)

**Thread distribution:** total ~105–112% CPU across 47 threads, but the **GUI/main thread alone is ~79% of one core**; every other thread (panadapter worker, audio, Qt pool) is ≤10%. Classic single-GUI-thread cap.

**Main-thread self-time by library:**

| Library | Self % | What |
|---|---|---|
| libQt6Gui | ~45% | `QImage` copy / fill / blend (raster compositing) |
| libnvidia-eglcore | ~29% | NVIDIA GL driver — present / composite / texture upload |
| libc | ~16% | `memcpy`/`memset` behind the QImage ops |
| Qt6Core / Qt6Widgets | ~4% | event dispatch |
| **AetherSDR (our code)** | **~2%** | — |

**Call graph (caller view):**

```
QApplicationPrivate::notify_helper (45%)
└─ QWidget::event (42.7%)
   ├─ QWidgetRepaintManager::paintAndFlush (31.8%)
   │  └─ flush → QPlatformBackingStore::rhiFlush (27.2%)
   │     ├─ QImage::detach() → QImage::copy(QRect const&)   ~15.2%   ← full backing-store stage copy / present
   │     └─ QRhi::endFrame → swapchain present              ~11.8%
   ├─ QRhiWidget::paintEvent → renderGpuFrame → QImage::fill(transparent)  ~7%  (appears ×2 → multiple panadapters)
   └─ QPushButton::paintEvent → QCommonStyle::drawControl   ~3.5%
```

## Interpretation

With **4 `QRhiWidget` panadapters in one top-level window** mixed with raster sibling widgets, Qt appears to repeatedly (a) copy/fill the raster backing store on the CPU via `QImage` (~61% across Qt6Gui + libc) and (b) push it through the NVIDIA GL composite/present (~29%) — **per panadapter present**. That CPU composite, not the GPU, is the 10–15 fps cap.

## Two candidate fixes (pending investigation)

1. **Reduce the CPU `QImage` backing-store recopy (~61%).** If Qt re-stages an *unchanged* full-window backing store on every RHI present (rather than dirty-rect / cached), guarding/caching that — or giving each panadapter its own native subsurface (`WA_NativeWindow`) / a single shared RHI surface so they bypass the shared raster backing-store composite — is the big win. **Open question: is `rhiFlush`'s copy full-window or dirty-rect, and is the staged texture cached across frames?**
2. **Backend swap to Vulkan (~29%).** Would attack the NVIDIA GL-driver present cost (and possibly some staging). Requires `setApi(Vulkan)` + a `QVulkanInstance` in `main.cpp`; not a launch-flag. May not touch the ~61% CPU `QImage` cost if that's backend-agnostic.

## Status

A scoped deep-research investigation into Qt 6.11's exact `rhiFlush` backing-store / dirty-rect behavior is **in flight** to decide between fix #1 and #2 (and rank the mitigations: `WA_NativeWindow` per panadapter, per-panadapter top-level `QWindow`, single full-window RHI surface, `WA_StaticContents`, QOpenGLWidget vs QRhiWidget, Vulkan backend). Findings will be posted as a comment.

Prior related perf work: GPU-composited slice flags (#3617, #3695), flag re-raster idle-pause (#3746/#3764).

**Comment by ten9876** [comment]
## Deep-research findings — Qt 6.11 `rhiFlush` backing-store behavior (verified against primary source)

A scoped investigation (61 claims verified against Qt 6.11.1 source `qbackingstoredefaultcompositor.cpp` / `qrhiwidget.cpp` / official docs + the Qt 6.4 graphics blog by RHI maintainer L. Agocs) resolves which fix is real. **Headline: my "fix #1 = cache the backing-store texture" framing was wrong — Qt already caches and dirty-rects. The real lever is *what dirties the backing store every frame* + giving the panadapters their own native surface. Vulkan (fix #2) is demoted — it can't touch the dominant cost.**

### What `rhiFlush` actually does (verified, not full-window)

`QPlatformBackingStore::rhiFlush()` → `QBackingStoreDefaultCompositor::toTexture()`:

- Keeps a **cached** `mutable std::unique_ptr<QRhiTexture> m_texture`.
- **Early-out when nothing changed:** `if (dirtyRegion.isEmpty() && !resized) return m_texture.get();` — *no copy, no upload*.
- When dirty (not resized): uploads **only the dirty sub-rect** (`dirtyRegion.boundingRect() & imageRect` via `QRhiTextureSubresourceUploadDescription`), **not** the whole window.
- Full upload happens **only** on first frame or window resize.

So the `QImage::detach()` + sub-rect copy we profiled at ~15% **is not unavoidable per-frame work** — it only runs when the *raster backing store has a non-empty dirty region that frame*. **Conclusion: something is dirtying the raster layer every frame, defeating the early-out.** (The `image.detach()` is a COW ownership break that copies the whole image; `setSourceTopLeft/Size` "may trigger a QImage copy depending on format and backend" — verbatim from the docs.)

### The compositing model (verified)

- Any `QRhiWidget`/`QOpenGLWidget`/`QQuickWidget` flips the **entire top-level window** to GPU-composited mode. **All** QPainter (raster) widget content in that window "now needs to be uploaded into a texture as well" to be composited (Qt 6.4 blog, verbatim).
- Multiple GPU widgets do **not** each trigger a separate full flush: the compositor holds `m_textureQuadData` (one `PerQuadData` per GPU widget) + one `m_widgetQuadData` for the raster store, and does **one** `beginPass` into **one** swapchain, drawing one quad per widget. So 4 panadapters = **4 quads in 1 shared composite**, not 4 full flushes. The 4× is in **flush *frequency*** (the union of all widgets' update requests), not per-flush cost.
- The compositor is **backend-agnostic QRhi** since 6.4 ("same set of draw calls… with Vulkan, Metal, or Direct 3D"). **The raster-to-texture staging copy lives in QtGui and is NOT a GL artifact** → **Vulkan would not remove the ~61% CPU `QImage` cost**; it would only change the ~29% NVIDIA driver-side present/upload.

### Refuted (do not pursue)

- ❌ "rhiFlush re-uploads the whole window every frame" — false; cached + dirty-rect.
- ❌ `WA_OpaquePaintEvent` / `autoFillBackground=false` / `WA_StaticContents` as fixes — **refuted as overreach**; they don't touch the composite/staging path.
- ❌ QTBUG-45425 — it's a 16-bit-backing-store *correctness* bug, not a perf bug. Not relevant.

### Ranked fixes for our "4 QRhiWidgets in one raster window" topology

1. **Eliminate the per-frame raster-backing-store dirtying (kills the ~15% detach + its upload).** Find the raster sibling repainting every frame so Qt's `dirtyRegion.isEmpty()` early-out fires on panadapter-only frames. Prime suspects: an animating raster child (SmartMTR meter ballistics, a status/clock widget, the dB scale). **Cheapest, highest-leverage on the CPU cost.** → next action: profile *what* marks the backing store dirty each frame.
2. **Give each panadapter its own native surface** — `WA_NativeWindow` on the `QRhiWidget`, or `QWindow`+`QWidget::createWindowContainer()`. Qt docs, verbatim: createWindowContainer is "faster… due to the **lack of the additional composition step**." This takes the GPU content out of the shared backing-store composite entirely. **Highest-leverage on the composite cost**, more invasive.
3. **Coalesce / cap panadapter update frequency** — cuts flush frequency, scaling down both the ~61% and ~29% proportionally.
4. **Vulkan backend** — only addresses the ~29% NVIDIA-driver slice; leaves the ~61% CPU `QImage` cost intact. **Demoted** from the original framing.

### Verdict on the issue's fix #1 vs #2

**Fix #1 is the real one**, but via *dirty-region suppression + native surface*, not texture caching (Qt already caches). **Fix #2 (Vulkan) is a minor, separate win** capped at the ~29% driver cost. Next concrete step: instrument which raster widget dirties the backing store every frame (point 1), since that's the cheapest large win and validates the whole model.

<sub>Synthesised from the deep-research run (97 agents, 61 verified claims); the workflow's final formatting step errored, so this was reconstructed from the verified-claim journal. Primary sources: qtbase `qbackingstoredefaultcompositor.cpp` @6.11.1, `qrhiwidget.cpp`, doc.qt.io QRhiWidget/QOpenGLWidget/QRhiTextureSubresourceUploadDescription, qt.io "Qt Quick and Widgets, Qt 6.4 Edition".</sub>

**Comment by svabi79** [comment]
@ten9876 — a Windows/D3D11 reference data point that may help triangulate fix #1,
FWIW. Your deep-research conclusion (rhiFlush is cached + dirty-rect; the lever is
*what dirties the backing store every frame*; the composite is backend-agnostic;
Vulkan demoted) lines up with what I measure here — with one extra angle: a
platform that does **not** cap.

## Setup
- **Windows 11, D3D11** (QRhiWidget default; log: `backend: D3D11`), DWM compositor.
- **i9-13900K (24C/32T) + desktop RTX 4090** (single dGPU) — comparable class to your
  i9-13980HX + RTX 4090 Laptop, so this isn't raw-HW headroom masking the cost; and
  since the bottleneck is single-GUI-thread-bound (your point), the extra cores don't
  help anyway.
- **4 panadapters**, AetherSDR window **~4860×1400 device-px (~6.8 MP backing
  store)** on a 5120×1440 49" panel, **100% scaling** (dpr 1).
- Same `WA_NativeWindow`-less, shared-raster-backing-store path as Linux (the
  macOS native-surface branch is `#ifdef Q_OS_MAC`).

## Measurement (idle, mouse off the panes; aether.perf)
| | 1 panadapter | 4 panadapters |
|---|---|---|
| process CPU | 0.67 cores | **0.91 cores** (≈ +0.08/pan) |
| panAgeP95 (display latency) | ~14 ms | **~11 ms** (no growth) |
| fps | ~29 | **~26/pan** (aggregate ~104) |
| renderP95 | 0.3 ms | 0.2 ms |

**No cap** — 4 panadapters stay smooth and low-latency.

## Why this corroborates fix #1 (and where it points)
- Windows runs the **same shared raster backing store, no native surface** as Linux —
  yet doesn't cap. So the cap isn't the backing-store *mechanism* (Windows composites
  the 4 quads through it fine); it's the **per-frame *dirtying*** that defeats
  `dirtyRegion.isEmpty()` — exactly your fix #1.
- **Resolution makes this stronger, not weaker:** the staging copy scales with the
  dirtied pixel area, and this window (~6.8 MP) is larger than a laptop panel. A
  full-window per-frame dirty would cap *harder* at this resolution — it doesn't. So
  on Windows the backing store is genuinely **not** dirtied per panadapter frame.
- **Localizing the dirty source:** the cross-platform raster children you flagged
  (SmartMTR ballistics, dB scale) run + animate identically here and still don't cap
  Windows. So the per-frame dirty looks **Linux/Wayland-specific** (a platform
  backing-store-dirty behavior) rather than a cross-platform widget repaint — which
  may narrow the "profile *what* marks the store dirty" hunt.

## Honest caveats
Whole-stack difference (Windows/DWM/D3D11/clean-dGPU vs Linux/Wayland/OpenGL/hybrid),
so it isolates "the Windows path doesn't dirty per-frame," not a single variable. And
it's a *negative* result (no repro) — its value is as the control that confirms the
mechanism itself isn't the cap.

Happy to run targeted Windows-side instrumentation if useful — e.g. log what marks
the backing store dirty on panadapter-only frames here, for a direct Linux contrast.

**Comment by jensenpat** [comment]
Closed in #3958
